Cleaned up material definitions+updated trackers to compare against v15

cjbarton151 opened the pull request:
I am updating _ldmx-sw_, here are the details.

In v14, geometry definitions are kind of all over the place in more than one sense. They're located in various files, inconsistently named, and often re-defined. Materials exported from other software may even have slightly different material properties, like density, composition, or atomic mass.

Compounding this issue, many python scripts used for validation, tracking etc. are still pointing to 'ldmx-det-v14-8gev' or 'ldmx-det-v14-8gev-nocals'. I don't know if v15 is stable enough for all of the python scripts yet, but using a script which references v14 but then uses v15 causes ldmx-sw to load both geometries and then apply one. However, the material definitions are still loaded, causing all the warnings.

### What are the issues that this addresses?
<!--
_Hint_: Use the phrase '_This resolves #< issue number >_' so that they are linked automatically.
-->

This (mostly) resolves #1892 . Read below if you want to know why it doesn't fully resolve it, or you can skip the next paragraph.

I don't consider this fully resolved for 3 reasons.
1) As you can see in the screenshots, the detector is still set twice in the ecal_pn test (I haven't run the other validation tests manually but I assume there will be similar behavior). This is related to the next point,
2) There is still exactly _one_ re-definition of each material. On the bright side, this is kind of nice because it lists all of the defined materials in the terminal output during runtime - but is still unintended. I wasted some time trying to track (pun intended) down this issue, and it seems to have something to do with the tracking, specifically `geo.py` and/or `full_tracking_sequence.py`. I don't want to deal with this (or at best deal with it in a separate issue) so I'm not investigating further.
3) There are other python scripts in the repo that are still hardcoded to use v14 instead of v15. Again, I don't feel like re-validating all of these scripts with v15, and that should be a separate issue anyway.

## Check List
- [X] I successfully compiled _ldmx-sw_ with my developments.
- [X] I read, understood and follow the [coding rules](https://ldmx-software.github.io/developing/coding-rules.html).
<!-- If these coding rules are confusing or you need help, still open the PR as a _draft_ and we can help you! -->
- [X] I ran my developments and the following shows that they are successful.
<!-- put plots or some other proof that your developments work and do the intended function -->

Before:
<img width="1355" height="763" alt="image" src="https://github.com/user-attachments/assets/cd8944cf-c207-4eed-9250-e51f8f545e08" />
<img width="897" height="562" alt="image" src="https://github.com/user-attachments/assets/80a4e99e-30b1-4f2a-92dc-4c40cb743a95" />

After:
<img width="1537" height="743" alt="image" src="https://github.com/user-attachments/assets/91a411bd-6614-4a47-a6b6-9f5d02f2900b" />

Commit:
Cleaned up material definitions. Updated a couple python scripts hard-coded to compare against v14 to compare against v15 instead.

Changed a ~ to a -.

tvami commented on Tracking/python/geo.py:
```diff
@@ -46,7 +46,7 @@ def __init__(self):
         else: 
             super().__init__('TrackersTrackingGeometry', 'tracking::geo::TrackersTrackingGeometryProvider', 'Tracking')
-            self.setDetector('ldmx-det-v14-8gev-no-cals')
+            self.setDetector('ldmx-det-v15-8gev')
```
Let's keep it as it was, the point of this is that the tracker doenst need to see the calorimeters for the tracking geometry, so it made Matt's studies much faster, in the end we should make a `ldmx-det-v15-8gev-no-cals` but that's only the relevant when the tracker geometry will change

tvami commented on Detectors/data/ldmx-det-v15-8gev/hcal.gdml:
```diff
@@ -240,7 +126,7 @@
       <!-- - - - - - - - -  FULL HCAL STRUCTURES - - - - - - - -  -->
     <volume name="hadronic_calorimeter">
-      <materialref ref="G4_AIR" />
+      <materialref ref="Air" />
```
I dont know what the difference is, did you move away from the G4 default on purpose?

cjbarton151 replied:
The GDML geometry builder doesn't have access to the Geant4 default materials intrinsically*
That material is named G4_AIR but is defined in GDML the same way every other material is:
<img width="495" height="200" alt="image" src="https://github.com/user-attachments/assets/70050e0d-698f-443b-b666-d9781f973b25" />

Most likely scenario: someone made that segment of the HCal in a different G4 simulation, then exported it. The other definition of "Air" (which I just took from one of the other files) doesn't include the trace carbon, and the other parameters are different by about a factor of 0.1%


*although there may be a way I don't know about to invoke G4NistManager::FindOrBuildMaterial from a GDML material definition 🤔

tvami commented on Detectors/data/ldmx-det-v15-8gev/ecal_motherboard5_assembly.gdml:
```diff
@@ -275,7 +275,7 @@
   <structure>
     <volume name="nohole_motherboard5_assembly">
-      <materialref ref="FR40x27264f0"/>
+      <materialref ref="FR4"/>
```
Is this just a simple rename? I'm worried if the FR4 existed somewhere else and the `FR40x27264f0` was done on purpose to make sure that it's unique. But it was before me, so maybe @tomeichlersmith can chime in

cjbarton151 replied:
This again looks like a material which was exported from another program. Ironically, that automatic naming convention on export is precisely to prevent duplicate material definitions... But this was the only place FR4 was defined in v14.

tomeichlersmith replied:
Can confirm, the auto-naming added this suffix and it was not left in out of anything besides laziness on my part.

tvami commented:
OK did you see we have a segfault?
```
 Setting detector for tracking to ldmx-det-v15-8gev
  
   *** Break *** segmentation violation
  
  
  
  ===========================================================
  There was a crash.
  This is the entire stack trace of all threads:
  ===========================================================
  #0  0x00007f8bcd2a07a7 in __GI___wait4 (pid=31, stat_loc=stat_loc
  entry=0x7ffd90c9cf28, options=options
  entry=0, usage=usage
  entry=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
  #1  0x00007f8bcd2a08eb in __GI___waitpid (pid=<optimized out>, stat_loc=stat_loc
  entry=0x7ffd90c9cf28, options=options
  entry=0) at ./posix/waitpid.c:38
  #2  0x00007f8bcd1e85bb in do_system (line=<optimized out>) at ../sysdeps/posix/system.c:172
  #3  0x00007f8bcd983583 in TUnixSystem::StackTrace() () from /usr/local/lib/libCore.so
  #4  0x00007f8bcd982ef4 in TUnixSystem::DispatchSignals(ESignals) () from /usr/local/lib/libCore.so
  #5  <signal handler called>
  #6  0x00007f8bcaa36cdd in simcore::geo::GDMLParser::read() () from /home/runner/work/ldmx-sw/ldmx-sw/install/lib/libSimCore_Geo.so
  #7  0x00007f8bcaa972a2 in simcore::SimulatorBase::buildGeometry() () from /home/runner/work/ldmx-sw/ldmx-sw/install/lib/libSimCore.so
  #8  0x00007f8bcaa98a5f in simcore::SimulatorBase::configure(framework::config::Parameters&) () from /home/runner/work/ldmx-sw/ldmx-sw/install/lib/libSimCore.so
  #9  0x00007f8bcdc7b34c in framework::Process::Process(framework::config::Parameters const&) () from /home/runner/work/ldmx-sw/ldmx-sw/install/lib/libFramework.so
  #10 0x0000555ef63cf0b9 in main ()
```

tomeichlersmith reviewed (commented):
I think this looks good - one thing I'm worried about is that we often use the material and/or volume name within the biasing and filtering code to check if we are within a certain region of the detector. In a better world, we would actually using Geant4 regions more and update the biasing/filtering code, but that requires a lot of validation that folks have been avoiding.

Have you checked if the biasing/filtering looks for any of these material names? (I don't think it does, I think it looks at volumes but I'm uncertain now.)

Commit:
Updated hardcoded material dependencies in the source code to account for new material naming conventions. Reverted the change to Tracking/python/geo.py, since v15 doesn't have a no-cals version (reverted v15->v14-no-cals).

Added backwards compatibility for <v15.

Re-added static attribute to tungsten_material. Added additional warning message for clarity when using old tungsten material.

tvami commented on SimCore/src/SimCore/BiasOperators/PhotoNuclear.cxx:
```diff
@@ -70,7 +70,7 @@ G4VBiasingOperation* PhotoNuclear::ProposeOccurenceBiasingOperation(
                  pn_xsec_unbiased_);
     static auto material_tungsten =
-        simcore::g4user::ptrretrieval::getMaterial("G4_W");
+        simcore::g4user::ptrretrieval::getMaterial("Tungsten");
```
With this change we loose backward compatibility, in this sense that we wont be able to run v14 geom anymore. Maybe rename Tungsten in v14 too? Or somehow make this code geometry version dependent? Probably easier to just rename W in v14 too

cjbarton151 replied:
Good catch. I made `material_tungsten` non-static and now, if PtrRetrieval can't find "Tungsten", it looks for "G4_W". This should keep it backwards compatible for all versions which use G4_W. I did similar for the other changes (involving G4_Si and G4_AIR).

tvami reviewed (approved):
Looks good! I'm not sure if we had a reason for the material to be `static` earlier but I think this should be good now

tomeichlersmith commented on SimCore/src/SimCore/BiasOperators/PhotoNuclear.cxx:
```diff
@@ -69,8 +69,11 @@ G4VBiasingOperation* PhotoNuclear::ProposeOccurenceBiasingOperation(
+        simcore::g4user::ptrretrieval::getMaterial("Tungsten");
+    // For versions < v15
+    if (material_tungsten == nullptr)
+      material_tungsten = simcore::g4user::ptrretrieval::getMaterial("G4_W");
```
Please leave the `static` in the declaration. The `getMaterial` function is slow since it has to do a string comparison across the materials loaded into memory but the materials do not change after the geometry is loaded so we only need to look up this pointer once and then we can use the pointer for the rest of the run.

**Context**
https://github.com/LDMX-Software/ldmx-sw/pull/1633
https://github.com/LDMX-Software/ldmx-sw/issues/1376

cjbarton151 replied:
No problem, I reverted it

tomeichlersmith commented on Detectors/data/ldmx-det-v15-8gev/materials.gdml:
```diff
@@ -1,6 +1,36 @@
+names that are coloquially understandable, like how we
+use "glue" instead of listing the chemical formula.
+
+-->
```
this is awesome, thank you for such detailed work!!!

tomeichlersmith reviewed (commented):
Excellent work, just requesting to put the `static` back in for potential-performance related reasons.

github-actions[bot] committed:
Apply clang-format and clang-tidy
